PR for module 3 project exercise

gmoraita opened the pull request:
Please review the Module 3 project exercise. All 3 Parts are included and updated README to provide additional instructions for the testing

gmoraita-dev committed:
Ignore workspace file

gmoraita-dev committed:
- Added ViewModel
- introduced a favicon.ico

gmoraita-dev committed:
Step 2 - completed. Added logic and tests for seggregating status items

gmoraita-dev committed:
Completed Part 1 step 3
- Added recent/older Done items logic and test cases
- Amended index accordingly

gmoraita-dev committed:
Optimised tester - no reliance on todoapi
Board can now be initialised with statuses

gmoraita-dev committed:
Step 3 - Autogenerate Items and minor optimisations

gmoraita-dev committed:
- Optimised ViewModel constructor and variables.
- Simplified test setup

gmoraita-dev committed:
Completed Part 3 and optimised code

gmoraita-dev committed:
Updated the README with latest info on testing

gmoraita-dev committed:
Refined README and minor comment removal in test_requests

gmoraita-dev committed:
Improved formatting for README

hugh-emerson commented on todo_app/data/todoapi.py:
```diff
@@ -9,34 +9,49 @@ class TrelloAPI():
+            return self.load_board_by_name(self.board_name)
+
+    def load_board_by_name(self, board_name) -> Board:
         boards = self.call_api('/1/members/%s/boards' % self.user, 'GET', {'lists':'all'}).json()
```
You can actually make a request to `.../members/me/...` instead of configuring a user id.

gmoraita replied:
But that will return the boardId whereas here I want to search by board name. Which saves me having to do a second lookup

hugh-emerson commented on todo_app/tests_e2e/test_e2e.py:
```diff
@@ -0,0 +1,67 @@
+import pytest
```
You should have installed pytest and selenium with e.g. `poetry add pytest` which would also add it to pyproject.toml for you. Or better `poetry add pytest selenium --dev`.
Currently people who run this project would have to manually install those packages themselves in order to run the tests.

And then you should be running pytest with `poetry run pytest`.

Instructions for running the tests should also be added to the README.md

gmoraita replied:
Added poetry add pytest selenium selinium-requests mock --dev. Also updated the README

hugh-emerson commented on todo_app/tests_e2e/test_e2e.py:
```diff
@@ -0,0 +1,67 @@
+    assert driver.title == 'To-Do App'
+
+    #check item added
+    add_response = driver.request('POST', _TODO_URL+ '/add', data = {Item.item_title: 'a test I do', Item.item_due_date: '01/02/2021', Item.item_description: 'a test description'})
+    assert add_response.text.count('a test I do') == 1
```
selenium tests are UI tests. You should be using the webpage as a real user would e.g. by simulating entering text into inputs and clicking buttons.

You can use `driver` to find elements on the page and interact with them.

gmoraita replied:
Done - using Driver to interact with the page as end user

hugh-emerson commented on todo_app/tests_e2e/test_e2e.py:
```diff
@@ -0,0 +1,67 @@
+
+    #check item deleted
+    delete_response = driver.request('POST', _TODO_URL+'/delete/%s' % new_task_id)
+    assert delete_response.text.count('a test I do') == 0
```
Similarly you shouldn't assert things about the HTTP response directly, you should assert what is displayed in the browser

gmoraita replied:
Done - using Driver to get elements from the page.

hugh-emerson commented on todo_app/tests_e2e/test_e2e.py:
```diff
@@ -0,0 +1,67 @@
+
+def create_temp_board():
+    todoapi = TrelloAPI()
+    board = todoapi.call_api('/1/boards/', 'POST', {'name': _TEST_BOARD_NAME+'_'+date.today().strftime(r"%Y%m%d") + datetime.now().strftime(r"%H%M%S%f")}).json()
```
A datetime is a date and time, so you can use just the datetime to build the whole timestamp:
`datetime.now().strftime(r"%Y%m%d%H%M%S%f")`

gmoraita replied:
Good point - thanks. I changed.

hugh-emerson commented on todo_app/tests/test_requests.py:
```diff
@@ -0,0 +1,43 @@
+mockresponses = MockResponses()
+
+
+def mock_get_lists(method, url, **kwargs):
```
The function name is misleading

gmoraita replied:
Changed to mock_responses to reflect the returned of mocked responses

hugh-emerson commented on .env.test:
```diff
@@ -0,0 +1,10 @@
+
+FLASK_SKIP_DOTENV=True
+
+trello_user=mytrellouser
```
This is unused and should be deleted.

gmoraita replied:
Done - updated the file with variables needed

hugh-emerson commented on todo_app/tests/test_app.py:
```diff
@@ -0,0 +1,76 @@
+import pytest
```
The filename is misleading. These are unit tests for your view model. So this should be called something like `test_view_model.py`

gmoraita replied:
done

hugh-emerson commented on todo_app/tests/mock_responses.py:
```diff
@@ -0,0 +1,991 @@
+
+
+    @property
+    def trello_user(self):
+        return 'bobinaki'
```
This "trello_user" is not part of the mock response... it's part of your app config. And it's not just test config, it actually has to match the real (and uncommitted) value in trello.config in order for the test to pass. Which is not great.
Similarly for the board_name has to match the real board name in trello.config because your app checks that `board.name == board_name:`

Separate to getting the tests to pass, you currently read the board_id from an environment variable but everything else from your trello.config file, which is confusing and causes extra complication.

The exercise instructions explain how to override the config when running tests but assumes you were using environment variables. With a trello.config file, things become more complicated.

You should either switch to just using environment variables or if you feel attached to the trello.config file, you can keep it but things get a bit more complicated:
You should have multiple versions of it, where the test version is added to git. Then the test should tell TrelloAPI to use the test one, either via a single env var or even by overriding TrelloAPI.CREDS_CONFIG_FILE, but that can cause other issues.
And in later modules you will need to do a little extra work to create the config file when running it in a container or on a server.

gmoraita replied:
Removed completely dependency on trello.config and deleted it. All now across the project, is dependenct on environment vars based on .env / .env.test

hugh-emerson reviewed (changes requested):
Mostly there but the tests only pass on your machine or with some undocumented fiddling because:
- A few packages are missing from pyproject.toml. See my comment on test_e2e.py
- The `test_requests` test is actually dependent on values in your local trello.config file. See my comment on mock_responses.py

I've also left some other minor comments.

gmoraita-dev committed:
Implemented all feedback from PR:
- Using now env instead of the trelloapi.config. As part of this, nw use app_config.py
- selenium driver used for simulation of user actions and doing assertions
- separated the call to trello api into a new class to streamline re-use between normal use and testing
- updated README

hugh-emerson commented on pyproject.toml:
```diff
@@ -10,8 +10,16 @@ flask = "^1.1.2"
 python-dotenv = "^0.14.0"
 python-dateutil = "2.7.3"
 requests = "2.22.0"
+pytest = "^6.2.1"
+selenium = "^3.141.0"
+selenium-requests = "^1.3"
+mock = "^4.0.3"
```
You can remove these lines now that they're included in dev-dependencies.

hugh-emerson reviewed (approved):
Looks good! Merge when you're ready.

gmoraita-dev committed:
Removed obsolete dependencies
